src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -588,47 +592,153 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
 }
 
 /*
- Return value is a "Win32 BOOL" (1 = true, 0 = false)
- */
-int32_t GlobalizationNative_StartsWith(
-                        SortHandle* pSortHandle,
-                        const UChar* lpTarget,
-                        int32_t cwTargetLength,
-                        const UChar* lpSource,
-                        int32_t cwSourceLength,
-                        int32_t options)
+ collation element is an int used for sorting. It consists of 3 components:
+    * primary - first 16 bits, representing the base letter
+    * secondary - next 8 bits, typically an accent
+    * tertiary - last 8 bits, typically the case
+
+An example (the numbers are made up to keep it simple)
+    a: 1 0 0
+    ą: 1 1 0
+    A: 1 0 1
+    Ą: 1 1 1
+
+    this method returns a mask that allows for characters comparison using specified Collator Strength
+*/
+static int32_t GetCollationElementMask(UColAttributeValue strength)
 {
-    int32_t result = FALSE;
-    UErrorCode err = U_ZERO_ERROR;
-    const UCollator* pColl = GetCollatorFromSortHandle(pSortHandle, options, &err);
+    switch (strength)
+    {
+        case UCOL_PRIMARY:
+            return UCOL_PRIMARYORDERMASK;
+        case UCOL_SECONDARY:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK;
+        default:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK | UCOL_TERTIARYORDERMASK;
+    }
+}
 
-    if (U_SUCCESS(err))
+static int32_t inline GetNextNonIgnorableCharacter(UCollationElements* pIterator, UErrorCode* pErrorCode)
+{
+    int32_t result = UCOL_IGNORABLE;
+
+    do
     {
-        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, &err);
-        int32_t idx = USEARCH_DONE;
+        result = ucol_next(pIterator, pErrorCode);
+    }
+    while (result == UCOL_IGNORABLE); // we don't check errorCode because on error the result is set to UCOL_NULLORDER
```

Comment:
>while (result == UCOL_IGNORABLE); // we don't check errorCode because on error the result is set to UCOL_NULLORDER [](start = 4, length = 114)

not checking the error code can cause a weird behavior if it happen. this method will return UCOL_NULLORDER which can make SimpleStartsWith_Iterators return true and that is wrong. may be you just OR all error codes and then check it one time at the end?

Reply:
very good point, I am going to add a check for that  to the calling method